README.md
# php-panaceamobile

An API client for sending SMSs via the [Panacea Mobile](https://www.panaceamobile.com) API

[![Author](http://img.shields.io/badge/author-@superbalist-blue.svg?style=flat-square)](https://twitter.com/superbalist)
[![Build Status](https://img.shields.io/travis/Superbalist/php-panaceamobile/master.svg?style=flat-square)](https://travis-ci.org/Superbalist/php-panaceamobile)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE)
[![Packagist Version](https://img.shields.io/packagist/v/superbalist/php-panaceamobile.svg?style=flat-square)](https://packagist.org/packages/superbalist/php-panaceamobile)
[![Total Downloads](https://img.shields.io/packagist/dt/superbalist/php-panaceamobile.svg?style=flat-square)](https://packagist.org/packages/superbalist/php-panaceamobile)


## Installation

```bash
composer require superbalist/php-panaceamobile
```

## Integrations

Want to get started quickly? Check out some of these integrations:

* Laravel - https://github.com/Superbalist/simple-sms-panacea-mobile

## Usage

```php
$username = 'your_username';
$password = 'your_password';

$guzzleClient = new \GuzzleHttp\Client();

$client = new \Superbalist\PanaceaMobile\PanaceaMobileAPI($guzzleClient, $username, $password);

$response = $client->sendMessage('+27000000000', 'This is a test message.');
var_dump($response);
```


changelog.md
# Changelog

## 1.0.0 - 2016-10-07

* Initial release

tests/PanaceaMobileAPITest.php
<?php

namespace Tests;

use GuzzleHttp\Client;
use GuzzleHttp\Psr7\Request;
use Mockery;
use PHPUnit\Framework\TestCase;
use Superbalist\PanaceaMobile\PanaceaMobileAPI;

class PanaceaMobileAPITest extends TestCase
{
    public function testSetGetUri()
    {
        $guzzleClient = Mockery::mock(Client::class);
        $client = new PanaceaMobileAPI($guzzleClient);
        $this->assertEquals('http://api.panaceamobile.com', $client->getUri());
        $client->setUri('http://127.0.0.1');
        $this->assertEquals('http://127.0.0.1', $client->getUri());
    }

    public function testSetGetUsername()
    {
        $guzzleClient = Mockery::mock(Client::class);
        $client = new PanaceaMobileAPI($guzzleClient, 'my_username');
        $this->assertEquals('my_username', $client->getUsername());
        $client->setUsername('my_new_username');
        $this->assertEquals('my_new_username', $client->getUsername());
    }

    public function testSetGetPassword()
    {
        $guzzleClient = Mockery::mock(Client::class);
        $client = new PanaceaMobileAPI($guzzleClient, 'my_username', 'my_password');
        $this->assertEquals('my_password', $client->getPassword());
        $client->setPassword('my_new_password');
        $this->assertEquals('my_new_password', $client->getPassword());
    }

    public function testSetGetUserAgent()
    {
        $guzzleClient = Mockery::mock(Client::class);
        $client = new PanaceaMobileAPI($guzzleClient);
        $this->assertEquals('superbalist/php-panaceamobile', $client->getUserAgent());
        $client->setUserAgent('lorem-ipsum');
        $this->assertEquals('lorem-ipsum', $client->getUserAgent());
    }

    public function testMakeBaseUri()
    {
        $guzzleClient = Mockery::mock(Client::class);
        $client = new PanaceaMobileAPI($guzzleClient);
        $client->setUri('http://api.panaceamobile.com//');
        $uri = $client->makeBaseUri('json');
        $this->assertEquals('http://api.panaceamobile.com/json', $uri);
    }

    public function testGetGlobalHeaders()
    {
        $guzzleClient = Mockery::mock(Client::class);
        $client = new PanaceaMobileAPI($guzzleClient);
        $client->setUserAgent('My UserAgent String');
        $headers = $client->getGlobalHeaders();
        $this->assertArrayHasKey('User-Agent', $headers);
        $this->assertEquals('My UserAgent String', $headers['User-Agent']);
    }

    public function testGet()
    {
        $request = new Request(
            'GET',
            'http://api.panaceamobile.com/json?action=message_send',
            ['User-Agent' => 'superbalist/php-panaceamobile']
        );

        $guzzleClient = Mockery::mock(Client::class);
        $client = Mockery::mock(
            '\Superbalist\PanaceaMobile\PanaceaMobileAPI[createRequest,sendRequest]',
            [$guzzleClient]
        );
        $client->shouldAllowMockingProtectedMethods();
        $client->shouldReceive('createRequest')
            ->withArgs([
                'GET',
                'json?action=message_send',
            ])
            ->once()
            ->andReturn($request);
        $client->shouldreceive('sendRequest')
            ->with($request)
            ->once();

        $client->get('message_send');
    }

    public function testGetWithQueryString()
    {
        $request = new Request(
            'GET',
            'http://api.panaceamobile.com/json?action=message_send',
            ['User-Agent' => 'superbalist/php-panaceamobile']
        );

        $guzzleClient = Mockery::mock(Client::class);
        $client = Mockery::mock(
            '\Superbalist\PanaceaMobile\PanaceaMobileAPI[createRequest,sendRequest]',
            [$guzzleClient]
        );
        $client->shouldAllowMockingProtectedMethods();
        $client->shouldReceive('createRequest')
            ->withArgs([
                'GET',
                'json?hello=world&action=message_send',
            ])
            ->once()
            ->andReturn($request);
        $client->shouldreceive('sendRequest')
            ->with($request)
            ->once();

        $client->get('message_send', ['hello' => 'world']);
    }

    public function testSendMessage()
    {
        $guzzleClient = Mockery::mock(Client::class);
        $client = Mockery::mock('\Superbalist\PanaceaMobile\PanaceaMobileAPI[get]', [$guzzleClient]);
        $client->shouldReceive('get')
            ->withArgs([
                'message_send',
                [
                    'to' => '+27000000000',
                    'text' => 'This is my message',
                    'from' => '+27111111111',
                    'report_mask' => 19,
                    'report_url' => null,
                    'charset' => null,
                    'data_coding' => null,
                    'message_class' => -1,
                    'auto_detect_encoding' => 0,
                ]
            ])
            ->once();
        $client->sendMessage('+27000000000', 'This is my message', '+27111111111');
    }
}


src/PanaceaMobileAPI.php
<?php

namespace Superbalist\PanaceaMobile;

use GuzzleHttp\Client;
use GuzzleHttp\Psr7\Request;
use Psr\Http\Message\RequestInterface;

class PanaceaMobileAPI
{
    /**
     * @var Client
     */
    protected $client;

    /**
     * @var string
     */
    protected $uri = 'http://api.panaceamobile.com';

    /**
     * @var string
     */
    protected $userAgent = 'superbalist/php-panaceamobile';

    /**
     * @var string
     */
    protected $username;

    /**
     * @var string
     */
    protected $password;

    /**
     * @param Client $client
     * @param string $username
     * @param string $password
     */
    public function __construct(Client $client, $username = null, $password = null)
    {
        $this->client = $client;
        $this->username = $username;
        $this->password = $password;
    }

    /**
     * Set the uri.
     *
     * @param string $uri
     */
    public function setUri($uri)
    {
        $this->uri = $uri;
    }
    /**
     * Return the uri.
     *
     * @return string
     */
    public function getUri()
    {
        return $this->uri;
    }

    /**
     * Set the username.
     *
     * @param string $username
     */
    public function setUsername($username)
    {
        $this->username = $username;
    }

    /**
     * Return the username.
     *
     * @return string
     */
    public function getUsername()
    {
        return $this->username;
    }

    /**
     * Set the password.
     *
     * @param string $password
     */
    public function setPassword($password)
    {
        $this->password = $password;
    }

    /**
     * Return the password.
     *
     * @return string
     */
    public function getPassword()
    {
        return $this->password;
    }

    /**
     * Set the user-agent HTTP header.
     *
     * @param string $userAgent
     */
    public function setUserAgent($userAgent)
    {
        $this->userAgent = $userAgent;
    }
    /**
     * Return the user-agent HTTP header.
     *
     * @return string
     */
    public function getUserAgent()
    {
        return $this->userAgent;
    }

    /**
     * Return the full uri to an API end-point.
     *
     * @param string $endpoint
     * @return string
     */
    public function makeBaseUri($endpoint)
    {
        return rtrim($this->uri, '/') . '/' . trim($endpoint, '/');
    }

    /**
     * Return an array of headers to send with every HTTP request.
     *
     * @return array
     */
    public function getGlobalHeaders()
    {
        return [
            'User-Agent' => $this->userAgent,
        ];
    }

    /**
     * Create an HTTP request object.
     *
     * @param string $method
     * @param string $endpoint
     * @param mixed $body
     * @param array $headers
     * @return Request
     */
    protected function createRequest($method, $endpoint, $body = null, array $headers = [])
    {
        $uri = $this->makeBaseUri($endpoint);
        $headers = array_merge($this->getGlobalHeaders(), $headers);
        return new Request($method, $uri, $headers, $body);
    }

    /**
     * Make an HTTP GET request.
     *
     * **Example**
     *
     * ```php
     * $response = $client->get('message_send', ['to' => '+27000000000', 'text' => 'Hello World', 'from' => '+27000000000']);
     * ```
     *
     * @param string $action
     * @param array $query
     * @return mixed
     */
    public function get($action, array $query = [])
    {
        $query['action'] = $action;
        $query['username'] = $this->username;
        $query['password'] = $this->password;

        $endpoint = 'json';

        if (!empty($query)) {
            $endpoint .= '?' . http_build_query($query);
        }

        $request = $this->createRequest('GET', $endpoint);
        return $this->sendRequest($request);
    }

    /**
     * Send an HTTP request.
     *
     * @param RequestInterface $request
     * @return array
     * @throws \Exception
     */
    protected function sendRequest(RequestInterface $request)
    {
        $response = $this->client->send($request);
        $json = json_decode($response->getBody(), true);

        if (isset($json['status']) && $json['status'] !== 1) {
            throw new \Exception($json['details']);
        }

        return $json;
    }

    /**
     * Send an SMS.
     *
     * **Example**
     *
     * ```php
     * $index = $client->sendMessage('+27000000000', 'This is my message content!');
     * ```
     *
     * @param string $to
     * @param string $message
     * @param string $from
     * @param int $reportMask
     * @param string $reportUrl
     * @param string $charset
     * @param int $dataCoding
     * @param int $messageClass
     * @param bool $autoDetectEncoding
     * @return mixed
     */
    public function sendMessage(
        $to,
        $message,
        $from = null,
        $reportMask = 19,
        $reportUrl = null,
        $charset = null,
        $dataCoding = null,
        $messageClass = -1,
        $autoDetectEncoding = false
    ) {
        $data = [
            'to' => $to,
            'text' => $message,
            'from' => $from,
            'report_mask' => $reportMask,
            'report_url' => $reportUrl,
            'charset' => $charset,
            'data_coding' => $dataCoding,
            'message_class' => $messageClass,
            'auto_detect_encoding' => $autoDetectEncoding ? 1 : 0,
        ];
        return $this->get('message_send', $data);
    }
}
